### PR TITLE
feat: [CDS-46655]: List Active Services NG

### PR DESCRIPTION
**Pull request by ivan-83**
## Describe your changes
We are getting the list of active services by using the following queries stated on the wiki page. Firstly, we get the list of active services and after that update the service name, org ad project names. 
Here, we are using java JDBC and will try to rewrite not complex query parts by using JOOQ in the next PR related to filtering.   

@harness-david  Could you pls review the following queries when you get a chance? Thanks in advance!
https://harness.atlassian.net/wiki/spaces/CDP/pages/21239005811/Tech+Spec+Active+Services+Service+Licenses#SQL-Queries

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41072)

### CODE REVIEW

**Comment by ivan-83** [comment]
trigger ti1

**Comment by ivan-83** [comment]
trigger ti4

File: 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java [hunk 1]
```diff
@@ -135,4 +156,32 @@ public class LicenseUsageResource {
...
+  @Path("cd/active-services")
+  @ApiOperation(value = "List Active Services in CD Module", nickname = "lisCDActiveServices")
+  @Operation(operationId = "lisCDActiveServices",
+      summary =
+          "List Active Services with instances, last deployed and licenses consumed details on Account, Ord and Project level",
+      responses =
+      { @io.swagger.v3.oas.annotations.responses.ApiResponse(description = "Returns a list of active services") })
+  @NGAccessControlCheck(resourceType = "LICENSE", permission = "core_license_view")
+  public ResponseDTO<Page<LicenseUsageDTO>>
+  lisCDActiveServices(@Parameter(description = ACCOUNT_PARAM_MESSAGE) @QueryParam(
```
Comment (neelam-harness) [inline 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java:170]: I think we mean `lisCDActiveServices` -> `listCDActiveServices` ?

File: 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+public class PageableUtils {
+  public Pageable getPageRequest(int page, int size, List<String> sort, Sort sortBy) {
+    try {
+      if (isEmpty(sort)) {
+        return PageRequest.of(page, size, sortBy);
+      } else {
+        return getPageRequest(page, size, sort);
+      }
+    } catch (Exception e) {
+      throw new InvalidRequestException(e.getMessage(), e);
```
Comment (neelam-harness) [inline 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java:37]: Let's log the exception wherever applicable before throwing ?

File: 960-ng-license-usage-beans/src/main/java/io/harness/licensing/usage/beans/cd/CDLicenseUsageConstants.java [hunk 3]
```diff
@@ -19,4 +21,13 @@ public class CDLicenseUsageConstants {
   public static final int TIME_PERIOD_IN_DAYS = 30;
   public static final double PERCENTILE = 0.95;
   public static final int SERVICE_INSTANCE_LIMIT = 20;
+  public static final String DEFAULT_FILTER_PROPERTIES_VALUE = "all";
+  public static final String SERVICE_INSTANCES_QUERY_PROPERTY = "serviceInstances";
+  public static final String SERVICE_INSTANCES_SORT_PROPERTY = "instanceCount";
+  public static final String LAST_DEPLOYED_SERVICE_QUERY_PROPERTY = "lastDeployed";
+  public static final String LAST_DEPLOYED_SERVICE_SORT_PROPERTY = "lastDeployed";
```
Comment (neelam-harness) [inline 960-ng-license-usage-beans/src/main/java/io/harness/licensing/usage/beans/cd/CDLicenseUsageConstants.java:28]: nit: curious if `LAST_DEPLOYED_SERVICE_QUERY_PROPERTY` and `LAST_DEPLOYED_SERVICE_SORT_PROPERTY` are both same ? If so, can we name the property more generic and re-use the same string ?

File: 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    return org.springframework.data.domain.PageRequest.of(page, size, Sort.by(orders));
+  }
+
+  public static void validateSort(Sort sort, List<String> allowedSortProperties) {
+    for (Sort.Order next : sort) {
+      String property = next.getProperty();
+      if (!allowedSortProperties.contains(property)) {
+        throw new InvalidArgumentsException(format("Invalid sort property: %s", property));
```
Comment (neelam-harness) [inline 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java:65]: let's log the exception before throwing ?

**Review by neelam-harness — APPROVED** [review]
Overall looks good to me. Thanks for doing this, Ivan!

Minor nit:  noticed that we don't have Logging anywhere. Let's add logs wherever applicable before submitting.

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41072)* status: 22 file(s) to review. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/usage/impl/CDLicenseUsageImpl.java` line 106 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3EO-r2-106:-NIgQjel2BmCHqCT_3tX:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/125-cd-nextgen/src/main/java/io/harness/cdng/usage/impl/CDLicenseUsageImpl.java#L106)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Let's add Logger to log the erroneous cases if any such as bad arguments etc. Also logging the success cases might be helpful in the scenarios where we might need troubleshooting.
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/usage/impl/CDLicenseUsageImpl.java` line 143 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3EO-r2-143:-NIgRU9NDBCbK3agU3ec:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/125-cd-nextgen/src/main/java/io/harness/cdng/usage/impl/CDLicenseUsageImpl.java#L143)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

let's add a TODO here and the jira ticket to track this for later ?
</blockquote></details>

Done.
___
*[`945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java` line 163 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3ES-r2-163:-NIgOot21XgplX_GdHD4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java#L163)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

did we mean `list`CDActiveServices here ?
</blockquote></details>

Done.
___
*[`945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java` line 165 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3ES-r2-165:-NIgOshQ2OQRkA87kHrJ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java#L165)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

let's replace `Ord` -> `Organization` ?
</blockquote></details>

Done.
___
*[`945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java` line 170 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3ES-r2-170:-NIgP0TVAGXg6HFW1k4J:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java#L170)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

I think we mean `lisCDActiveServices` -> `listCDActiveServices` ?
</blockquote></details>

Done.
___
*[`945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java` line 37 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3Eg-r2-37:-NIgPHp_247tnMV7b730:bukvc0y) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java#L37)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Let's log the exception wherever applicable before throwing ?
</blockquote></details>

We have a centralized exception handler where we are logging all exceptions thrown by the app, and here we are using the principle "throw and log" to prevent duplicate logs messages. 
___
*[`945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/41072#-NIUkZ-tiZNshhusR3Eg-r2-65:-NIgPxBWC5wruzm6j32H:bf26u1) ([raw file](https://github.com/harness/harness-core/blob/c146e97ce9b18437a26c0d5d32a1e08791e8d835/945-license-usage-sdk/src/main/java/io/harness/licensing/usage/utils/PageableUtils.java#L65)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

let's log the exception before throwing ?
</blockquote></details>

The same as above.

**Review by deepak-harness — APPROVED** [review]
> Reviewed 11 of 21 files at r1, 6 of 8 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41072)* status: 22 file(s) to review. 7. unresolved discussions. 1 of 2 approvals obtained.

File: 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java [hunk 5]
```diff
@@ -135,4 +160,33 @@ public class LicenseUsageResource {
         (ServiceInstanceUsageDTO) licenseUsageInterface.getLicenseUsage(accountIdentifier, ModuleType.CD, timestamp,
             CDUsageRequestParams.builder().cdLicenseType(CDLicenseType.SERVICE_INSTANCES).build()));
   }
+
+  @POST
+  @Path("cd/active-services")
+  @ApiOperation(value = "List Active Services in CD Module", nickname = "lisCDActiveServices")
```
Comment (neelam-harness) [inline 945-license-usage-sdk/src/main/java/io/harness/licensing/usage/resources/LicenseUsageResource.java:166]: nit: rename `lisCDActiveServices` -> `listCDActiveServices`

**Review by neelam-harness — APPROVED** [review]

**Review by neelam-harness — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41072#-:-NIhZeHh8HVkfaTWjfOQ:bnfp4nl)* status: 22 file(s) to review.

**Review by neelam-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41072#-:-NIhZjdz0K_lDqhz__zA:bnfp4nl)* status: 22 file(s) to review.

**Review by vladimir-peric — APPROVED** [review]
> Reviewed 11 of 21 files at r1, 6 of 8 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41072)* status: :shipit: complete! ✓

**Pull request merged**